File: EPI/10.1/typescript/10.1.ts
export interface Node {
    left?: Node;
    right?: Node;
}

export function isBalanced(root: Node): boolean {
    return isBalancedRec(root).balanced;
}

function isBalancedRec(root: Node | undefined): {balanced: boolean, height: number} {
    if (!root) {
        return {balanced: true, height: 0};
    }

    const {balanced: lb, height: lh} = isBalancedRec(root.left);
    const {balanced: rb, height: rh} = isBalancedRec(root.right);

    return {
        balanced: lb && rb && Math.abs(lh - rh) <= 1,
        height: Math.max(lh, rh) + 1
    };
}

File: EPI/10.1/swift/10.1.swift
// swiftlint:disable conditional_binding_cascade

public class Node {
    public let left: Node?
    public let right: Node?

    public init(left: Node? = nil, right: Node? = nil) {
        self.left = left
        self.right = right
    }

    public func balanced() -> Bool {
        if case .Balanced = status(self) {
            return true
        }
        return false
    }
}

enum Status {
    case Unbalanced
    case Balanced(height: Int)
}

func status(_ tree: Node?) -> Status {
    if let t = tree {
        if case let .Balanced(leftHeight) = status(t.left),
           case let .Balanced(rightHeight) = status(t.right),
           abs(leftHeight - rightHeight) <= 1 {
            return .Balanced(height: 1 + max(leftHeight, rightHeight))
        }
        return .Unbalanced
    }
    return .Balanced(height: -1)
}
